libevmasm/SemanticInformation.cpp

```diff
@@ -232,6 +232,47 @@ bool SemanticInformation::movable(Instruction _instruction)
 	return true;
 }
 
+bool SemanticInformation::movableIfStateInvariant(Instruction _instruction)
```

Comment:
This function returns `false` for instructions that are not movable but also do not read from or write to the state - is that how we want to design it?

Reply:
Are there instructions that read from the state, but do not write to the state and it would still be a problem to move them around?

Reply:
> This function returns false for instructions that are not movable but also do not read from or write to the state - is that how we want to design it?`

What do you think is an alternative?

> Are there instructions that read from the state, but do not write to the state and it would still be a problem to move them around?

These are the list of instructions that do not get moved because they have side effects:

STOP, KECCAK256, CALLDATACOPY, CODECOPY, EXTCODECOPY, RETURNDATACOPY, MLOAD, MSTORE, MSTORE8, SSTORE, JUMP, JUMPI, JUMPDEST, LOG0, LOG1, LOG2, LOG3, LOG4, CREATE, CALL, CALLCODE, RETURN, DELEGATECALL, STATICCALL, CREATE2, REVERT, INVALID, SELFDESTRUCT

These instructions are explicitly forbidden in `SemanticInformation::movable()`

KECCAK256, BALANCE, SELFBALANCE, EXTCODESIZE, EXTCODEHASH, RETURNDATASIZE, SLOAD, PC, MSIZE, GAS

Among these, I don't think there are instructions left that will qualify.